controllers/virtualmachinesnapshot/virtualmachinesnapshot_controller.go

```diff
@@ -158,3 +184,213 @@ func (r *Reconciler) ReconcileNormal(ctx *pkgctx.VirtualMachineSnapshotContext)
+			} else if errors.Is(err, virtualmachine.ErrVMSnapshotNotFound) {
+				ctx.Logger.Info("vmSnapshot not found in VC, continue reconciling")
+			} else {
+				return err
+			}
+		}
+
+		parent, err := r.updateParentSnapshot(ctx)
+		if err != nil {
+			return err
+		}
+
+		if err := r.updateVMStatus(ctx, parent); err != nil {
+			return err
+		}
+
+		controllerutil.RemoveFinalizer(vmSnapshot, Finalizer)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) deleteSnapshot(ctx *pkgctx.VirtualMachineSnapshotContext) error {
+	vmSnapshot := ctx.VirtualMachineSnapshot
+	// TODO: set removeChildren and consolidate to false by default for now.
+	if err := r.VMProvider.DeleteSnapshot(ctx.Context, vmSnapshot, ctx.VM, false, nil); err != nil {
+		return fmt.Errorf("failed to delete snapshot: %w", err)
+	}
+	return nil
+}
+
+func (r *Reconciler) updateParentSnapshot(ctx *pkgctx.VirtualMachineSnapshotContext) (*vmopv1.VirtualMachineSnapshot, error) {
+	ctx.Logger.V(5).Info("Updating parent snapshot")
+	vmSnapshot := ctx.VirtualMachineSnapshot
+	parent, err := r.getParentSnapshot(ctx, vmSnapshot)
+	if err != nil {
+		if errors.Is(err, virtualmachine.ErrParentVMSnapshotNotFound) {
+			ctx.Logger.V(5).Info("parent snapshot not found")
+			return nil, nil
+		}
+		return nil, fmt.Errorf("failed to get parent snapshot: %w", err)
+	}
+	ctx.Logger.V(5).Info("parent snapshot found", "parent", parent.Name)
+
+	// remove current snapshot from parent's children
+	parentPatch := client.MergeFrom(parent.DeepCopy())
+	for i, child := range parent.Status.Children {
+		if child.Name == vmSnapshot.Name {
+			parent.Status.Children = slices.Delete(parent.Status.Children, i, i+1)
+			break
+		}
+	}
+	children := vmSnapshot.Status.Children
+	if children != nil {
+		ctx.Logger.V(5).Info("Add children snapshots of current snapshot to parent's children")
+		// merge current's children to parent's children.
+		// make sure no duplicates are added.
+		for _, child := range children {
+			if !slices.Contains(parent.Status.Children, child) {
+				parent.Status.Children = append(parent.Status.Children, child)
+			}
+		}
+	}
+	if err := r.Status().Patch(ctx, parent, parentPatch); err != nil {
+		return nil, fmt.Errorf("failed to patch parent snapshot %s with children: %w", parent.Name, err)
+	}
+	return parent, nil
+}
+
+// getParentSnapshotName returns the parent snapshot name of the given snapshot.
+// It first checks if the snapshot is a root snapshot, if not, it will traverse the root snapshots to find the parent snapshot.
+// The upper limit of snapshot count of a VM is 128. The overhead here could be considered acceptable.
+func (r *Reconciler) getParentSnapshot(ctx *pkgctx.VirtualMachineSnapshotContext, vmSnapshot *vmopv1.VirtualMachineSnapshot) (*vmopv1.VirtualMachineSnapshot, error) {
+	rootSnapshots := ctx.VM.Status.RootSnapshots
+	if len(rootSnapshots) == 0 {
+		ctx.Logger.V(5).Info("no root snapshots found for VM")
+		return nil, virtualmachine.ErrParentVMSnapshotNotFound
```

Comment:
I'm not a fan of importing this package for its error type. For how this error is being used in its callstack I think we can define its own error type private to the controller.